src/sknnr_spatial/image.py

```diff
@@ -0,0 +1,346 @@
+
+        mask = np.zeros(self.flat_array.shape, dtype=bool)
+        flat_image = flat_image.astype(np.float64)
+
+        # If it's floating point, always mask NaNs
+        if is_float:
+            mask |= np.isnan(self.flat_array)
+
+        # If NoData was specified, mask those values
+        if self.nodata_vals is not None:
+            mask |= self.flat_array == self.nodata_vals
+
+        # Set the mask where any band contains NoData
+        flat_image[mask.max(axis=-1)] = np.nan
+
+        return flat_image
+
+    def _preprocess(self, array: NDArray, nan_fill: float = 0.0) -> NDArray:
+        """Preprocess the chunk by flattening to (pixels, bands) and filling NaNs."""
+        flat = array.reshape(-1, array.shape[-1])
+        if nan_fill is not None:
+            flat[np.isnan(flat)] = nan_fill
+
+        return flat
+
+    def _postprocess(self, array: NDArray, mask_nodata: bool = True) -> NDArray:
+        """Postprocess the chunk by unflattening to (y, x, band) and masking NoData."""
+        output_shape = [*self.array.shape[:2], -1]
+        if mask_nodata:
+            array = self._mask_nodata(array)
+
+        return array.reshape(output_shape)
+
+    def apply(
+        self, func, returns_tuple=False, mask_nodata=True, **kwargs
+    ) -> NDArray | tuple[NDArray]:
+        """
+        Apply a function to the flattened, processed chunk.
+
+        The function should accept and return one or more NDArrays in shape
+        (pixels, bands). The output will be reshaped back to the original chunk shape.
+        """
+        flat_result = func(self.flat_array, **kwargs)
+
+        if returns_tuple:
+            return tuple(
+                self._postprocess(result, mask_nodata=mask_nodata)
+                for result in flat_result
+            )
+
+        return self._postprocess(flat_result, mask_nodata=mask_nodata)
+
+
+class Image(Generic[ImageType], ABC):
+    """A wrapper around a multi-band image"""
+
+    band_dim_name: str
+    band_dim: int
+    band_names: NDArray
+
+    def __init__(self, image: ImageType, nodata_vals: NoDataType = None):
+        self.image = image
+        self.n_bands = self.image.shape[self.band_dim]
+        self.nodata_vals = self._validate_nodata_vals(nodata_vals)
+
+    def _validate_nodata_vals(self, nodata_vals: NoDataType) -> NDArray | None:
+        """
+        Get an array of NoData values in the shape (bands,) based on user input.
+
+        Scalars are broadcast to all bands while sequences are checked against the
+        number of bands and cast to ndarrays. There is no need to specify np.nan as a
+        NoData value because it will be masked automatically for floating point images.
+        """
+        if nodata_vals is None:
+            return None
+
+        # If it's a numeric scalar, broadcast it to all bands
+        if isinstance(nodata_vals, (float, int)) and not isinstance(nodata_vals, bool):
+            return np.full((self.n_bands,), nodata_vals)
```

Comment:
Good point. I don't think it would ever cause issues since Numpy will happily compare equality for floats and ints, but it's definitely inconsistent. 

Maybe instead of casting the `np.full` to `float` I should just not specify `dtype` for sequences? That way we'll just let Numpy figure it out based on the value(s) provided by the user? I'm not sure why I specified `float` originally...

Reply:
Seems like a reasonable solution.